src/Bitmap_cubical_complex/include/gudhi/Persistence_on_rectangle.h

```diff
@@ -0,0 +1,595 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s): Marc Glisse
+ *
+ *    Copyright (C) 2023 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+//  ds_find_set_ is inspired from code in Boost.Graph that is
+//
+//  (C) Copyright Jeremy Siek 2004
+//  Distributed under the Boost Software License, Version 1.0. (See
+//  accompanying file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+
+#ifndef PERSISTENCE_ON_RECTANGLE_H
+#define PERSISTENCE_ON_RECTANGLE_H
+
+#include <gudhi/Debug_utils.h>
+#ifdef GUDHI_DETAILED_TIMES
+ #include <gudhi/Clock.h>
+#endif
+
+#include <boost/range/adaptor/reversed.hpp>
+
+#ifdef GUDHI_USE_TBB
+ #include <tbb/parallel_sort.h>
+#endif
+
+#ifdef DEBUG_TRACES
+ #include <iostream>
+#endif
+#include <vector>
+#include <memory>
+#include <algorithm>
+#include <stdexcept>
+#include <cstddef>
+
+namespace Gudhi::cubical_complex {
+
+// When building a cubical complex from top-dimensional cells, there are
+// normally more vertices than input top cells ((x+1)*(y+1) instead of x*y).
+// However, the top cells along the boundary turn out to be collapsible, and we
+// only need to work with (x-1)*(y-1) vertices.
+template <class Filtration_value, class Index = std::size_t, bool output_index = false>
+struct Persistence_on_rectangle {
```

Comment:
I guess, everything is public the time we decide on an interface?

Reply:
The class is really meant as an implementation detail, at least for now. I started with a single function, and the things that are now member functions were lambdas. But this was inconvenient to work with because tools don't handle lambdas so well, and the profile mostly said that most of the time was taken by operator() calling operator(), followed by operator()...
If I had to make something public right now, it would be the free function at the end, while the class would remain hidden.

Reply:
If only the free function is public, there is no possibility to only compute H0 or only H1. Does computing both not make much of a difference than computing only one? I am wondering because in one of your comments you write that `primal` and `dual` can be dominant.